src/clang_parser.cpp

```diff
@@ -274,13 +275,158 @@ CXErrorCode ClangParser::ClangParserHandler::parse_translation_unit(
+        {
+          std::string macro_name;
+          std::string macro_value;
+          if (translateMacro(c, macro_name, macro_value))
+          {
+            auto &macros = static_cast<BPFtrace*>(client_data)->macros_;
+            macros[macro_name] = macro_value;
+          }
+          return CXChildVisit_Recurse;
+        }
+
+        if (clang_getCursorKind(parent) == CXCursor_EnumDecl)
+        {
+          auto &enums = static_cast<BPFtrace*>(client_data)->enums_;
+          enums[get_clang_string(clang_getCursorSpelling(c))] = clang_getEnumConstantDeclValue(c);
+          return CXChildVisit_Recurse;
+        }
+
+        if (clang_getCursorKind(parent) != CXCursor_StructDecl &&
+            clang_getCursorKind(parent) != CXCursor_UnionDecl)
+          return CXChildVisit_Recurse;
+
+        if (clang_getCursorKind(c) == CXCursor_FieldDecl)
+        {
+          auto &structs = static_cast<BPFtrace*>(client_data)->structs_;
+
+          auto named_parent = get_named_parent(c);
+          auto ptype = clang_getCanonicalType(clang_getCursorType(named_parent));
+          auto ptypestr = get_clang_string(clang_getTypeSpelling(ptype));
+          auto ptypesize = clang_Type_getSizeOf(ptype);
+
+          auto ident = get_clang_string(clang_getCursorSpelling(c));
+          auto offset = clang_Type_getOffsetOf(ptype, ident.c_str()) / 8;
+          auto type = clang_getCanonicalType(clang_getCursorType(c));
+
+          auto struct_name = get_clang_string(clang_getCursorSpelling(named_parent));
+          if (struct_name == "")
+            struct_name = ptypestr;
+          remove_struct_union_prefix(struct_name);
+
+          // Warn if we already have the struct member defined and is
+          // different type and keep the current definition in place.
+          if (structs.count(struct_name) != 0 &&
+              structs[struct_name].fields.count(ident)  != 0 &&
+              structs[struct_name].fields[ident].offset != offset &&
+              structs[struct_name].fields[ident].type   != get_sized_type(type) &&
+              structs[struct_name].size                 != ptypesize)
+          {
+            std::cerr << "type mismatch for " << struct_name << "::" << ident << std::endl;
+          }
+          else
+          {
+            structs[struct_name].fields[ident].offset = offset;
+            structs[struct_name].fields[ident].type = get_sized_type(type);
+            structs[struct_name].size = ptypesize;
+          }
+        }
+
+        return CXChildVisit_Recurse;
+      },
+      &bpftrace);
+
+  // clang_visitChildren returns a non-zero value if the traversal
+  // was terminated by the visitor returning CXChildVisit_Break.
+  return err == 0;
+}
+
+bool ClangParser::parse_btf_definitions(BPFtrace &bpftrace)
+{
+  if (ast::Expression::resolve.size() == 0)
+    return true;
+
+  BTF btf;
+  if (!btf.has_data())
+    return true;
+
+  // We need to process every type separately, because BTF::c_def
```

Comment:
hm... from just briefly skimming and reading this comment, I think this is suboptimal.

It would be way more efficient to do btf_dump__new() once, then call  btf_dump__dump_type() for each type you need, then compile resulting header file with all the definitions (it won't have duplicates). Creating and discarding btf_dump object for every single instance of referenced type is quite wasteful.

Also, keep in mind that btf__find_by_name() does a linear search across all types, so it might be more efficient to first create  a set of type names you care about, then do single linear pass over all BTF types and dump those that match one of the name in a set. That way it's O(N) or O(NlogM) at most (if using sorted set) vs O(NM), where N is total number of BTF types, and M is number of types referenced from BPF program code.

Reply:
ok, will change